content/courses/ada-in-practice/chapters/raii.rst

```diff
@@ -0,0 +1,480 @@
+clients |mdash| would now call these new protected procedures:
+
+.. code-block:: ada
+
+    with Ada.Text_IO;  use Ada.Text_IO;
+
+    package body P is
+
+       protected Threadsafe is
+          procedure Operation_1;
+          procedure Operation_2;
+       private
+          State : Integer := 0;
+       end Threadsafe;
+
+       protected body Threadsafe is
+
+          procedure Operation_1 is
+          begin
+             State := State + 1;  -- for example...
+             Put_Line ("State is now " & State'Image);
+          end Operation_1;
+
+          procedure Operation_2 is
+          begin
+             State := State - 1;  -- for example...
+             Put_Line ("State is now " & State'Image);
+          end Operation_2;
+
+       end Threadsafe;
+
+       procedure Operation_1 is
+       begin
+          Threadsafe.Operation_1;
+       end Operation_1;
+
+       procedure Operation_2 is
+       begin
+          Threadsafe.Operation_2;
+       end Operation_2;
+
+       --  ...
+
+    end P;
+
+As a result, there can be no race conditions on :ada:`State`, and the source
+code is both simple and robust.
+
+We could put the PO in the package spec and then clients could call the
+protected object's operations directly, but changing all the clients could be
+expensive.
+
+However, this new design is not portable because the two :ada:`Threadsafe`
+protected procedure bodies both call a potentially blocking operation, in this
+case :ada:`Ada.Text_IO.Put_Line`. Erroneous execution is the result. It might
+work as intended when executed, or it might do something else, or, if detected,
+:ada:`Program_Error` will be raised. On a run-time library built on top of an
+operating system, it may work because the OS may provide thread locking
+mechanisms that the run-time library can use. In that case a blocking operation
+just suspends the caller thread's execution temporarily without releasing the
+PO lock. Although the blocking operation would allow some other caller task to
+be dispatched, no other caller could acquire that same PO lock, so race
+conditions are prevented within that PO. When the blocking operation returns,
+the protected procedure body can continue executing, still holding the lock.
+However, on a run-time library that does not use locks for mutual exclusion
+|mdash| it can use priorities, in particular |mdash| another caller to that
+same PO could access the enclosed variables while the first caller is blocked,
+thus breaking the mutually exclusive access guarantee.
+
+Calling an I/O operation is not all that strange here, and those are not the
+only potentially blocking operation defined by the language.
+
+Note that moving the calls to :ada:`Put_Line` out of the PO procedure bodies,
+back to the regular procedure bodies that call those PO procedures, would solve
+the portability problem but would not work functionally. There would be no
+guarantee that a call to :ada:`Put_Line` would immediately follow the execution
```

Comment:
Let's change the above to:
"guarantee that, during execution, the call to :ada:`Put_Line` would immediately follow the execution"